### DISCUSSION TITLE
Usage of crystal plasticity models in the context of global strain approach

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#18776] [Q&A General] [answered]
Dear all,
Further to my work on integrating the micro-structural model with a larger scale component models, progress was made earlier in enabling global strain approach to solve the micro-scale model. The link here <https://github.com/idaholab/moose/discussions/18380> can be referred for an understanding of the related work.  The material model however was not based upon a crystal plasticity considerations and so my recent work has focussed on selecting the appropriate finite strain CP models in MOOSE that can realistically model the constitutive behaviour at this scale. I fiddled around with "FiniteStrainCrystalPlasticity" based objects and get them to work along with other material modules supporting the strain periodicity implementation. I have not been successful in running the computation and the following error occurs during the stage where the material modules are referenced. 

<code><pre>
*** ERROR ***
One or more Material Properties were not supplied on block 0:
deformation_gradient
</code></pre>

The following input structure has been used.

<code><pre>
[Mesh]
  [generated_mesh]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 5
    ny = 5
    nz = 5
  []
  [cnode]
    type = ExtraNodesetGenerator
    coord = '0.0 0.0 0.0'
    new_boundary = 100
    input = generated_mesh
  []
[]

[Variables]
  [./u_x]
    block = 0
  [../]
  [./u_y]
    block = 0
  [../]
  [./u_z]
    block = 0
  [../]
  [./global_strain]
    order = SIXTH
    family = SCALAR
  [../]
[]

[AuxVariables]
  [./disp_x]
    block = 0
  [../]
  [./disp_y]
    block = 0
  [../]
  [./disp_z]
    block = 0
  [../]
  [./s00]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./s11]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./e00]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./e11]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
[]

[AuxKernels]
  [./disp_x]
    type = GlobalDisplacementAux
    variable = disp_x
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 0
    block = 0
  [../]
  [./disp_y]
    type = GlobalDisplacementAux
    variable = disp_y
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 1
    block = 0
  [../]
  [./disp_z]
    type = GlobalDisplacementAux
    variable = disp_z
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 2
    block = 0
  [../]
  [./s00]
    type = RankTwoAux
    variable = s00
    rank_two_tensor = stress
    index_i = 0
    index_j = 0
    block = 0
  [../]
  [./s11]
    type = RankTwoAux
    variable = s11
    rank_two_tensor = stress
    index_i = 1
    index_j = 1
    block = 0
  [../]
  [./e00]
    type = RankTwoAux
    variable = e00
    rank_two_tensor = total_strain
    index_i = 0
    index_j = 0
    block = 0
  [../]
  [./e11]
    type = RankTwoAux
    variable = e11
    rank_two_tensor = total_strain
    index_i = 1
    index_j = 1
    block = 0
  [../]
[]

[GlobalParams]
  displacements = 'u_x u_y u_z'
  block = 0
[]

[Kernels]
  [./TensorMechanics]
  [../]
[]

[ScalarKernels]
  [./global_strain]
    type = GlobalStrain
    variable = global_strain
    global_strain_uo = global_strain_uo
  [../]
[]

[BCs]
  [./Periodic]
    [./all]
      auto_direction = 'x y z'
      variable = ' u_x u_y u_z'
    [../]
  [../]

  # fix center point location
  [./centerfix_x]
    type = DirichletBC
    boundary = 100
    variable = u_x
    value = 0
  [../]
  [./centerfix_y]
    type = DirichletBC
    boundary = 100
    variable = u_y
    value = 0
  [../]
  [./centerfix_z]
    type = DirichletBC
    boundary = 100
    variable = u_z
    value = 0
  [../]
[]

[Materials]
  [./crysp]
    type = FiniteStrainCrystalPlasticity
    block = 0
    gtol = 1e-2
    slip_sys_file_name = input_slip_sys.txt
    nss = 12
    num_slip_sys_flowrate_props = 2 #Number of properties in a slip system
    flowprops = '1 4 0.001 0.1 5 8 0.001 0.1 9 12 0.001 0.1'
    hprops = '1.0 541.5 60.8 109.8 2.5'
    gprops = '1 4 60.8 5 8 60.8 9 12 60.8'
    tan_mod_type = exact
    gen_random_stress_flag = false
    maximum_substep_iteration = 2
  [../]
  [./elasticity_tensor]
    type = ComputeElasticityTensorCP
    block = 0
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
  [../]
  [./strain]
    type = ComputeSmallStrain
    global_strain = global_strain
    block = 0
  [../]
  [./global_strain]
    type = ComputeGlobalStrain
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    block = 0
  [../]
  [./stress]
    type = ComputeLinearElasticStress
    block = 0
  [../]
[]

[UserObjects]
  [./global_strain_uo]
    type = GlobalStrainUserObject
    applied_stress_tensor = '5e9 0 0 0 0 0'
    execute_on = 'Initial Linear Nonlinear'
  [../]
[]

[Postprocessors]
  [./l2err_e00]
    type = ElementL2Error
    variable = e00
    function = 0.07142857 #strain_xx = C1111/sigma_xx
  [../]
  [./l2err_e11]
    type = ElementL2Error
    variable = e11
    function = -0.07142857*0.33 #strain_yy = -nu*strain_xx
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = 'PJFNK'

  line_search = basic

  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm         31   preonly   lu      1'

  l_max_its = 30
  nl_max_its = 12

  nl_rel_tol = 1.0e-10
  num_steps = 1
  dt = 0.01
[]

[Outputs]
  exodus = true
[]
</code></pre>

I am pretty sure that i am missing certain modules here that are vital in bridging these entities together. Could anybody advise me on the right way to address the material behaviour for the problems one might enounter in crsytal plasticity simulations in the context of a RVE model solved by a global strain approach?
Kind regards,
Arun

**Comment by sapitts** [comment, marked as answer]
Hi Arun,
We'd recommend that you use the newer material-based [Compute Multiple Crystal Plasticity Stress](https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.html) system. The material constitutive model, [Crystal Plasticity Kalidini Update](https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.html) will replicate the functionality used in your input file. The `ComputeMultipleCrystalPlasticityStress` system is being actively developed while the `FiniteStrainCrystalPlasticity` is not.

With any of our MOOSE crystal plasticity classes, one must supply an incremental strain calculator class, usually `ComputeFiniteStrain`. The error you are seeing occurs because only the incremental strain calculator classes provide the deformation gradient, and `ComputeSmallStrain` is a total strain calculator class which does not compute a deformation gradient. Please see the regression test input file [here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_method_test.i) as an example.

I would also recommend the use of the [TensorMechanicsMaster action](https://mooseframework.inl.gov/syntax/Modules/TensorMechanics/Master/index.html), if it will work well with the `GlobalStrain`. @SudiptaBiswas  and @dschwen  would you please provide guidance here?

Best Regards,
Stephanie

**Reply by SudiptaBiswas** [reply]
`GlobalStrain` approach works with `TensormechanicsMasterAction`. Here is an example, https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/global_strain/global_strain_action.i.

**Reply by abarun22** [reply]
Hi Stephanie, Sudipta,
Thanks for the suggestions. I went further ahead with `ComputeMultipleCrystalPlasticityStress` and the kalidindi crystal plasticity model and also included `TensormechanicsMasterAction`. I've removed the `GlobalStrain` inside the master block just to make sure it does not mix up with `ComputeGlobalStrain`.  This looks quite reasonable, however throws the following error.
<pre><code>
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

The following material properties are declared on block 0 by multiple materials:
  Material Property             Material Objects
  mechanical_strain             strain stress_div_strain
  total_strain                  strain stress_div_strain
</pre></code>

Looks like mutiple materials are trying to define the strain properties, which might need some trimming here. Here is the updated input commands for your reference.
<pre><code>
[Mesh]
  [generated_mesh]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 5
    ny = 5
    nz = 5
  []
  [cnode]
    type = ExtraNodesetGenerator
    coord = '0.0 0.0 0.0'
    new_boundary = 100
    input = generated_mesh
  []
[]
[GlobalParams]
  displacements = 'u_x u_y u_z'
  block = 0
[]
[Variables]
  [./u_x]
    block = 0
  [../]
  [./u_y]
    block = 0
  [../]
  [./u_z]
    block = 0
  [../]
  [./global_strain]
    order = SIXTH
    family = SCALAR
  [../]
[]

[AuxVariables]
  [./disp_x]
    block = 0
  [../]
  [./disp_y]
    block = 0
  [../]
  [./disp_z]
    block = 0
  [../]
  [./s00]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./s11]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./e00]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./e11]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
[]

[AuxKernels]
  [./disp_x]
    type = GlobalDisplacementAux
    variable = disp_x
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 0
    block = 0
  [../]
  [./disp_y]
    type = GlobalDisplacementAux
    variable = disp_y
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 1
    block = 0
  [../]
  [./disp_z]
    type = GlobalDisplacementAux
    variable = disp_z
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    component = 2
    block = 0
  [../]
  [./s00]
    type = RankTwoAux
    variable = s00
    rank_two_tensor = stress
    index_i = 0
    index_j = 0
    block = 0
  [../]
  [./s11]
    type = RankTwoAux
    variable = s11
    rank_two_tensor = stress
    index_i = 1
    index_j = 1
    block = 0
  [../]
  [./e00]
    type = RankTwoAux
    variable = e00
    rank_two_tensor = total_strain
    index_i = 0
    index_j = 0
    block = 0
  [../]
  [./e11]
    type = RankTwoAux
    variable = e11
    rank_two_tensor = total_strain
    index_i = 1
    index_j = 1
    block = 0
  [../]
[]

[ScalarKernels]
  [./global_strain]
    type = GlobalStrain
    variable = global_strain
    global_strain_uo = global_strain_uo
  [../]
[]

[Modules]
  [./TensorMechanics]
    # Master action for generating the tensor mechanics kernels, variables,
    # strain calculation material, and the auxilliary system for visualization
    [./Master]
      [./stress_div]
        strain = SMALL
        add_variables = true
        global_strain = global_strain #global strain contribution
        generate_output = 'strain_xx strain_xy strain_yy stress_xx stress_xy
                           stress_yy vonmises_stress'
      [../]
    [../]
  [../]
[]

[BCs]
  [./Periodic]
    [./all]
      auto_direction = 'x y z'
      variable = ' u_x u_y u_z'
    [../]
  [../]

  [./centerfix_x]
    type = DirichletBC
    boundary = 100
    variable = u_x
    value = 0
  [../]
  [./centerfix_y]
    type = DirichletBC
    boundary = 100
    variable = u_y
    value = 0
  [../]
  [./centerfix_z]
    type = DirichletBC
    boundary = 100
    variable = u_z
    value = 0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeElasticityTensorConstantRotationCP
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
  [../]
  [./stress]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'trial_xtalpl'
    tan_mod_type = exact
  [../]
  [./trial_xtalpl]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    tan_mod_type = exact
  [../]
  [./strain]
    type = ComputeFiniteStrain
    block = 0
    displacements = 'u_x u_y u_z'
  [../]
  [./global_strain]
    type = ComputeGlobalStrain
    scalar_global_strain = global_strain
    global_strain_uo = global_strain_uo
    block = 0
  [../]
[]

[UserObjects]
  [./global_strain_uo]
    type = GlobalStrainUserObject
    applied_stress_tensor = '5e9 0 0 0 0 0'
    execute_on = 'Initial Linear Nonlinear'
  [../]
[]

[Postprocessors]
  [./l2err_e00]
    type = ElementL2Error
    variable = e00
    function = 0.07142857 #strain_xx = C1111/sigma_xx
  [../]
  [./l2err_e11]
    type = ElementL2Error
    variable = e11
    function = -0.07142857*0.33 #strain_yy = -nu*strain_xx
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = 'PJFNK'

  line_search = basic

  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm         31   preonly   lu      1'

  l_max_its = 30
  nl_max_its = 12

  nl_rel_tol = 1.0e-10
  num_steps = 1
  dt = 0.01
[]

[Outputs]
  exodus = true
[]
</pre></code>

Any suggestions here will be highly appreciated.
Kind regards,
Arun

**Reply by sapitts** [reply]
Hi Arun, 
Only a couple of quick modifications to your input file ought to be all that's required to get your simulation running:
1. Remove the entire strain block, because this class is created by the TensorMechanics master action
  ```
  [./strain]
    type = ComputeFiniteStrain
    block = 0
    displacements = 'u_x u_y u_z'
  [../]
```
2. Change the strain setting in the TensorMechanics master action from `strain = SMALL` to `strain = FINITE`. This modification will create the `ComputeFiniteStrain` class from 1. automatically

Best Regards,
Stephanie

**Reply by abarun22** [reply]
Hi Stephanie,
Thanks for the suggestion. That indeed worked. The computation however would not converge. Could be a trick thing to solve given that the original model (one without CPFE material models) runs OK.  Please find here the execution log where the computation could not seem to get past the first time step.

<pre><code>
(base) abalasub@freia013> ../../../submodel-opt -i sub_gs_uniaxial_CPFE.i | tee -a run.log


*** Warning, This code is deprecated and will be removed in future versions:
The parameter 'use_legacy_dirichlet_bc' is no longer valid.

All Dirichlet boundary conditions are preset by default.

Remove said parameter in main to remove this deprecation warning.
Stack frames: 11
0: libMesh::print_trace(std::ostream&)
1: void moose::internal::mooseDeprecatedStream<libMesh::BasicOStreamProxy<char, std::char_traits<char> >, char const (&) [62], char const (&) [59], char const (&) [26], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [37]>(libMesh::BasicOStreamProxy<char, std::char_traits<char> >&, bool, char const (&) [62], char const (&) [59], char const (&) [26], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [37])
2: MooseApp::MooseApp(InputParameters)
3: submodelTestApp::submodelTestApp(InputParameters)
4: /home/abalasub/Working/Projects/submodel_repo/test/lib/libsubmodel_test-opt.so.0(+0xb0d8) [0x7fb7d98170d8]
5: std::shared_ptr<MooseApp> buildApp<submodelTestApp>(InputParameters const&)
6: AppFactory::createShared(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters, ompi_communicator_t*)
7: AppFactory::createAppShared(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, char**, ompi_communicator_t*)
8: ../../../submodel-opt() [0x4023b2]
9: __libc_start_main
10: ../../../submodel-opt() [0x402658]


Framework Information:
MOOSE Version:           git commit 626b041 on 2021-09-05
LibMesh Version:         ffab5c914090ad625645faa1bb4251d6b46f1c26
PETSc Version:           3.15.1
SLEPc Version:           3.15.1
Current Time:            Fri Sep 10 15:50:38 2021
Executable Timestamp:    Thu Sep  9 12:15:47 2021

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   216
  Elems:                   125
  Num Subdomains:          1

Nonlinear System:
  Num DOFs:                654
  Num Local DOFs:          654
  Variables:               { "u_x" "u_y" "u_z" } "global_strain"
  Finite Element Types:    "LAGRANGE" "SCALAR"
  Approximation Orders:    "FIRST" "SIXTH"

Auxiliary System:
  Num DOFs:                2023
  Num Local DOFs:          2023
  Variables:               { "disp_x" "disp_y" "disp_z" } { "s00" "s11" "e00" "e11" } { "strain_xx" "strain_xy"
                             "strain_yy" "stress_xx" "stress_xy" "stress_yy" "vonmises_stress" }
  Finite Element Types:    "LAGRANGE" "MONOMIAL" "MONOMIAL"
  Approximation Orders:    "FIRST" "CONSTANT" "CONSTANT"

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             Preconditioned JFNK
  MOOSE Preconditioner:    SMP

LEGACY MODES ENABLED:
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To remove this message, set 'use_legacy_material_output' to false in this application. If there are gold output files that contain material property output for which output occurs on INITIAL, then these will generate diffs due to zero values being stored, and these tests should be re-golded.



*** Warning ***
/home/abalasub/Working/Projects/submodel_repo/problem/Multiapps/full_solve_multiapp_1/sub_gs_uniaxial_CPFE.i:211: unused parameter 'Materials/trial_xtalpl/tan_mod_type'


Time Step 0, time = 0

Postprocessor Values:
+----------------+----------------+----------------+
| time           | l2err_e00      | l2err_e11      |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+----------------+


Scalar Variable Values:
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | global_strain_0| global_strain_1| global_strain_2| global_strain_3| global_strain_4| global_strain_5|
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+


Time Step 1, time = 0.01, dt = 0.01
 0 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 1 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 2 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 3 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 4 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 5 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 6 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 7 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 8 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 9 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
10 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
11 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
12 Nonlinear |R| = 5.000000e+09
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 12
 Solve Did NOT Converge!
Aborting as solve did not converge

Time Step 1, time = 0.005, dt = 0.005
 0 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 1 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 2 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 3 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 4 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 5 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 6 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 7 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 8 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 9 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
10 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
11 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
12 Nonlinear |R| = 5.000000e+09
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 12
 Solve Did NOT Converge!
Aborting as solve did not converge

Time Step 1, time = 0.0025, dt = 0.0025
 0 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 1 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 2 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 3 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 4 Nonlinear |R| = 5.000000e+09
      0 Linear |R| = 5.000000e+09
      1 Linear |R| = 5.000000e+09
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 1
 5 Nonlinear |R| = 5.000000e+09
</pre></code>

The global strains are possibly not tied to other objects of the material model in a correct manner i suppose. Probably we are missing some thing here.
Kind regards,
Arun

**Reply by SudiptaBiswas** [reply]
Does the model converge without the global strain? The global strain approach was not tested with the crystal plasticity model before, we might need modifications regarding calculating global strain incrementally.

**Reply by abarun22** [reply]
Hi Sudipta,
Yes the computation runs OK without global strain approach. Here are the trimmed input commands sans the global strain modules.
As you said, we might need the modifications for incremental global strain. I really do not have much idea how long the implementation would take. Do you reckon if this can be done quickly?
<pre><code>
[Mesh]
  [generated_mesh]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 5
    ny = 5
    nz = 5
  []
  [cnode]
    type = ExtraNodesetGenerator
    coord = '0.0 0.0 0.0'
    new_boundary = 100
    input = generated_mesh
  []
[]
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  block = 0
[]

[AuxVariables]
  [./s00]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./s11]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./e00]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
  [./e11]
    order = CONSTANT
    family = MONOMIAL
    block = 0
  [../]
[]

[AuxKernels]
  [./s00]
    type = RankTwoAux
    variable = s00
    rank_two_tensor = stress
    index_i = 0
    index_j = 0
    block = 0
  [../]
  [./s11]
    type = RankTwoAux
    variable = s11
    rank_two_tensor = stress
    index_i = 1
    index_j = 1
    block = 0
  [../]
  [./e00]
    type = RankTwoAux
    variable = e00
    rank_two_tensor = total_strain
    index_i = 0
    index_j = 0
    block = 0
  [../]
  [./e11]
    type = RankTwoAux
    variable = e11
    rank_two_tensor = total_strain
    index_i = 1
    index_j = 1
    block = 0
  [../]
[]

[Modules]
  [./TensorMechanics]
    # Master action for generating the tensor mechanics kernels, variables,
    # strain calculation material, and the auxilliary system for visualization
    [./Master]
      [./stress_div]
        strain = FINITE
        add_variables = true
        displacements = 'disp_x disp_y disp_z'
        generate_output = 'strain_xx strain_xy strain_yy stress_xx stress_xy
                           stress_yy vonmises_stress'
      [../]
    [../]
  [../]
[]

[BCs]
  [./symmy]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  [../]
  [./symmx]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  [../]
  [./symmz]
    type = DirichletBC
    variable = disp_z
    boundary = back
    value = 0
  [../]
  [./tdisp]
    type = FunctionDirichletBC
    variable = disp_z
    boundary = front
    function = '0.01*t'
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeElasticityTensorConstantRotationCP
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
    euler_angle_1 = 120.0
    euler_angle_2 = 125.264
    euler_angle_3 =  45.0
  [../]

  [./stress]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'trial_xtalpl'
    tan_mod_type = exact
  [../]
  [./trial_xtalpl]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    tan_mod_type = exact
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = 'PJFNK'

  line_search = basic

  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm         31   preonly   lu      1'

  l_max_its = 30
  nl_max_its = 12

  nl_rel_tol = 1.0e-10
  num_steps = 5
  dt = 0.01
[]

[Outputs]
  exodus = true
[]
</pre></code>

**Reply by abarun22** [reply]
Hi Sudipta,
I am keen to drive this forward. At the moment i do not have much idea about the extent of modifications needed for the work (global strain in incremental setting) and so its hard for me to arrive at an approximate time estimate. You might be in a better position to address the requirements and come up with a time frame. Probably you could initiate proceedings and show me the path, i could then see if the implmentation can be taken up parallel to my current work.

**Reply by abarun22** [reply]
Hi Sudipta,
Time has come for me to work on this piece of activity again. What i need from you at this stage is some minimal advise on how to enable the strains calculated by the global strain modules in an incremental manner, so that it blends nicely with the other computation modules, working under a time stepping set up. 
I look forward to hear your response.
Thanks and regards,
Arun
